File: Advanced/Arrays/Logic Building/4_union_of_2_sorted_arrays.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>
#include <climits>
using namespace std;

// --- Brute Force Approach 1 ---
// TC = O(nlogn) + O(mlogm) + O((n+m) log(n+m))
// SC = O(n+m)
vector<int> unionArrayBrute1(const vector<int> &nums1, const vector<int> &nums2)
{
    set<int> s;
    for (int num : nums1)
        s.insert(num);
    for (int num : nums2)
        s.insert(num);

    vector<int> unionArr(s.begin(), s.end());
    return unionArr;
}

// --- Optimal Approach 1 (Two Pointer Technique) ---
// TC = O(n + m)
// SC = O(n + m)
vector<int> unionArrayOptimal1(const vector<int> &nums1, const vector<int> &nums2)
{
    int n = nums1.size(), m = nums2.size();
    int i = 0, j = 0;
    vector<int> uni;

    while (i < n && j < m)
    {
        if (nums1[i] <= nums2[j])
        {
            if (uni.empty() || uni.back() != nums1[i])
                uni.push_back(nums1[i]);

            if (nums1[i] == nums2[j])
                j++;
            i++;
        }
        else
        {
            if (uni.empty() || uni.back() != nums2[j])
                uni.push_back(nums2[j]);
            j++;
        }
    }

    while (i < n)
    {
        if (uni.empty() || uni.back() != nums1[i])
            uni.push_back(nums1[i]);
        i++;
    }

    while (j < m)
    {
        if (uni.empty() || uni.back() != nums2[j])
            uni.push_back(nums2[j]);
        j++;
    }

    return uni;
}

// --- Optimal Approach 2 (Single Loop with Sentinels) ---
// TC = O(n + m)
// SC = O(n + m)
vector<int> unionArrayOptimal2(const vector<int> &nums1, const vector<int> &nums2)
{
    int n = nums1.size(), m = nums2.size();
    int i = 0, j = 0;
    vector<int> uni;

    while (i < n || j < m)
    {
        int val1 = (i < n) ? nums1[i] : INT_MAX;
        int val2 = (j < m) ? nums2[j] : INT_MAX;
        int candidate;

        if (val1 < val2)
        {
            candidate = val1;
            i++;
        }
        else if (val2 < val1)
        {
            candidate = val2;
            j++;
        }
        else
        {
            candidate = val1;
            i++;
            j++;
        }

        if (candidate != INT_MAX)
        {
            if (uni.empty() || uni.back() != candidate)
                uni.push_back(candidate);
        }
    }

    return uni;
}

// --- Helper to Print Vector ---
void printVector(const vector<int> &v)
{
    for (int x : v)
        cout << x << " ";
    cout << endl;
}

// --- Main Function ---
int main()
{
    vector<int> nums1 = {1, 2, 4, 5, 6};
    vector<int> nums2 = {2, 3, 5, 7};

    cout << "Array 1: ";
    printVector(nums1);
    cout << "Array 2: ";
    printVector(nums2);
    cout << endl;

    cout << "Brute Force 1: ";
    printVector(unionArrayBrute1(nums1, nums2));

    cout << "Optimal 1 (Two Pointer): ";
    printVector(unionArrayOptimal1(nums1, nums2));

    cout << "Optimal 2 (Single Loop + Sentinels): ";
    printVector(unionArrayOptimal2(nums1, nums2));

    return 0;
}

File: Advanced/Arrays/Logic Building/4_union_of_2_sorted_arrays.py
# Union of 2 sorted array


# Brute Force Approach 1
# TC = O(n)+O(m)+O((n+m)log(n+m))
# SC = O(n+m)
def unionArrayBrute1(nums1, nums2):
    s = set()
    for num in nums1:
        s.add(num)
    for num in nums2:
        s.add(num)
    union = list(sorted(s))
    return union


# Brute Force Approach 2
# TC = O(n) + O(m) + O((n+m)log(n+m))
# SC = O(n+m)
def unionArrayBrute2(nums1, nums2):
    return sorted(list(set(nums1) | set(nums2)))


# Optimal Approach 1 (Classic Two Pointer Technique)
# TC = O(n+m)
# SC = O(n+m)
def unionArrayOptimal1(nums1, nums2):
    n = len(nums1)
    m = len(nums2)
    i, j = 0, 0
    union = []

    while i < n and j < m:
        if nums1[i] <= nums2[j]:

            if not union or union[-1] != nums1[i]:
                union.append(nums1[i])

            if nums1[i] == nums2[j]:
                j += 1
            i += 1
        else:  # nums1[i] > nums2[j]
            if not union or union[-1] != nums2[j]:
                union.append(nums2[j])
            j += 1

    while i < n:
        if not union or union[-1] != nums1[i]:
            union.append(nums1[i])
        i += 1

    while j < m:
        if not union or union[-1] != nums2[j]:
            union.append(nums2[j])
        j += 1

    return union


# Optimal Approach 2 (Single Loop with Sentinels)
# TC = O(n+m)
# SC = O(n+m)
def unionArrayOptimal2(nums1, nums2):
    n, m = len(nums1), len(nums2)
    i, j = 0, 0
    union = []

    # Single loop continues as long as at least one array has elements
    while i < n or j < m:
        # Use float("inf") as a sentinel when an array is exhausted.
        # This avoids an IndexError and simplifies the main comparison logic.
        val1 = nums1[i] if i < n else float("inf")
        val2 = nums2[j] if j < m else float("inf")
        candidate = 0

        if val1 < val2:
            candidate = val1
            i += 1
        elif val2 < val1:
            candidate = val2
            j += 1
        else:
            candidate = val1
            i += 1
            j += 1

        # The check for infinity is necessary to stop at the end
        if candidate != float("inf"):
            if not union or union[-1] != candidate:
                union.append(candidate)

    return union
